jerry-core/ecma/base/ecma-gc.c

```diff
@@ -304,63 +304,47 @@ static void
 ecma_gc_mark_container_object (ecma_object_t *object_p) /**< object */
 {
   ecma_extended_object_t *map_object_p = (ecma_extended_object_t *) object_p;
-  ecma_object_t *internal_obj_p = ECMA_GET_INTERNAL_VALUE_POINTER (ecma_object_t,
-                                                                   map_object_p->u.class_prop.u.value);
-  ecma_extended_object_t *internal_ext_p = (ecma_extended_object_t *) internal_obj_p;
+  ecma_collection_t *container_p = ECMA_GET_INTERNAL_VALUE_POINTER (ecma_collection_t,
+                                                                    map_object_p->u.class_prop.u.value);
+  lit_magic_string_id_t lit_id = map_object_p->u.class_prop.class_id;
 
-  ecma_gc_set_object_visited (internal_obj_p);
-
-  jmem_cpointer_t prop_iter_cp = internal_obj_p->u1.property_list_cp;
-
-#if ENABLED (JERRY_PROPRETY_HASHMAP)
-  if (prop_iter_cp != JMEM_CP_NULL)
+  if (lit_id == LIT_MAGIC_STRING_WEAKSET_UL)
   {
-    ecma_property_header_t *prop_iter_p = ECMA_GET_NON_NULL_POINTER (ecma_property_header_t,
-                                                                     prop_iter_cp);
-    if (prop_iter_p->types[0] == ECMA_PROPERTY_TYPE_HASHMAP)
-    {
-      prop_iter_cp = prop_iter_p->next_property_cp;
-    }
+    return;
   }
-#endif /* ENABLED (JERRY_PROPRETY_HASHMAP) */
 
-  while (prop_iter_cp != JMEM_CP_NULL)
-  {
-    ecma_property_header_t *prop_iter_p = ECMA_GET_NON_NULL_POINTER (ecma_property_header_t, prop_iter_cp);
-    JERRY_ASSERT (ECMA_PROPERTY_IS_PROPERTY_PAIR (prop_iter_p));
+  uint8_t entry_size = ecma_op_container_entry_size (lit_id);
+  uint32_t entry_count = ECMA_CONTAINER_ENTRY_COUNT (container_p);
+  ecma_value_t *start_p = ECMA_CONTAINER_START (container_p);
 
-    ecma_gc_mark_properties ((ecma_property_pair_t *) prop_iter_p);
+  for (uint32_t i = 0; i < entry_count; i+= entry_size)
+  {
+    ecma_value_t *entry_p = start_p + i;
 
-    if ((internal_ext_p->u.class_prop.extra_info & ECMA_CONTAINER_FLAGS_WEAK) == 0)
+    if (ecma_is_value_empty (*entry_p))
     {
-      ecma_property_pair_t *prop_pair_p = (ecma_property_pair_t *) prop_iter_p;
+      continue;
+    }
 
-      for (uint32_t i = 0; i < ECMA_PROPERTY_PAIR_ITEM_COUNT; i++)
+    if (lit_id == LIT_MAGIC_STRING_SET_UL || lit_id == LIT_MAGIC_STRING_MAP_UL)
```

Comment:
I'm not a fan of repeatedly doing these checks inside a for loop. What do you think about having a separate loop for the different container types? It would result in slightly larger code, but it would be more efficient performance wise.

Reply:
@dbatyai I've updated the PR according to your notices.